python/pyspark/sql/protobuf/functions.py

```diff
@@ -0,0 +1,196 @@
+#
+# Licensed to the Apache Software Foundation (ASF) under one or more
+# contributor license agreements.  See the NOTICE file distributed with
+# this work for additional information regarding copyright ownership.
+# The ASF licenses this file to You under the Apache License, Version 2.0
+# (the "License"); you may not use this file except in compliance with
+# the License.  You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+"""
+A collections of builtin protobuf functions
+"""
+
+
+from typing import Dict, Optional, TYPE_CHECKING
+from pyspark import SparkContext
+from pyspark.sql.column import Column, _to_java_column
+from pyspark.util import _print_missing_jar
+
+if TYPE_CHECKING:
+    from pyspark.sql._typing import ColumnOrName
+
+
+def from_protobuf(
+    data: "ColumnOrName",
+    descFilePath: str,
+    messageName: str,
+    options: Optional[Dict[str, str]] = None,
+) -> Column:
+    """
+    Converts a binary column of Protobuf format into its corresponding catalyst value.
+    The specified schema must match the read data, otherwise the behavior is undefined:
+    it may fail or return arbitrary result.
+    To deserialize the data with a compatible and evolved schema, the expected
+    Protobuf schema can be set via the option protobuf descriptor.
+
+    .. versionadded:: 3.4.0
+
+    Parameters
+    ----------
+    data : :class:`~pyspark.sql.Column` or str
+        the binary column.
+    descFilePath : str
+        the protobuf descriptor in Message GeneratedMessageV3 format.
+    messageName: str
+        the protobuf message name to look for in descriptorFile.
+    options : dict, optional
+        options to control how the Protobuf record is parsed.
+
+    Notes
+    -----
+    Protobuf functionality is provided as an pluggable external module since Spark 3.4.0.
+
+    Examples
+    --------
+    >>> from pyspark.sql import Row
```

Comment:
This example is more complex because it uses both `to_protobuf` & `from_protobuf()`. 
Majority of the use case would be for `from_protobuf()`. Is it worth simplifying with just `from_protobuf()` demo? We could have some protobuf records in a file.

Reply:
@rangadi to make from_protobuf() function? We require a protobuf message. I'm simply converting a dataframe to a protobuf message with to_protobuf. Otherwise, we need to write a protobuf dynamic message in Python that includes the protobuf python lib installation. not sure which is the simplest one.

Reply:
How about just initializing bytes?
In fact, it does not need to be a working example. It could even read from kafka. With current example, simplicity of using this function to process protobufs is lost. 

Most users of this function may not much about protobufs at all. All they would know is that their input needs this function.
But this optional.

Reply:
@rangadi functions.py is not showing examples but is executing each line and trying to assert with exact values, its doctests(unit tests for pyspark). if we put an example that does not work, the build will fail. functions.py is part of python tests. we can add examples like just from_protobuf from bytes and reading from Kafka in SPARK-40776 right? 
```
    >>> protobufDf.show(truncate=False)
    +----------------------------------------+
    |protobuf                                |
    +----------------------------------------+
    |[08 02 12 05 41 6C 69 63 65 18 90 D5 06]|
    +----------------------------------------+
```

Reply:
I see, agree. Examples as suggested in SPARK-40776 sounds great. Most users will read that. 
We can leave it as it is.